drivers/flash/nrf_qspi_nor.c

```diff
@@ -461,36 +508,129 @@ static int qspi_nrfx_configure(const struct device *dev)
 	qspi_fill_init_struct(&QSPIconfig);
 
 	nrfx_err_t res = nrfx_qspi_init(&QSPIconfig, qspi_handler, dev_data);
+	int ret = qspi_get_zephyr_ret_code(res);
+
+	if ((ret == 0)
+	    && (INST_0_QER != JESD216_DW15_QER_NONE)) {
+		/* Set QE to match transfer mode.  If not using quad
+		 * it's OK to leave QE set, but doing so prevents use
+		 * of WP#/RESET#/HOLD# which might be useful.
+		 *
+		 * Note build assert above ensures QER is S1B6.  Other
+		 * options require more logic.
+		 */
+
+		ret = qspi_rdsr(dev);
+		if (ret < 0) {
+			LOG_ERR("RDSR failed: %d", ret);
+			return ret;
+		}
 
-	if (res == NRFX_SUCCESS) {
-		/* If quad transfer was chosen - enable it now */
-		if ((qspi_is_used_write_quad_mode(QSPIconfig.prot_if.writeoc))
-		    || (qspi_is_used_read_quad_mode(QSPIconfig.prot_if.readoc))) {
+		uint8_t sr = (uint8_t)ret;
+		bool qe_value =
+			(qspi_write_is_quad(QSPIconfig.prot_if.writeoc))
+			|| (qspi_read_is_quad(QSPIconfig.prot_if.readoc));
+		const uint8_t qe_mask = BIT(6); /* only S1B6 */
+		bool qe_state = ((sr & qe_mask) != 0U);
 
-			/* WRITE ENABLE has to be sent before QUAR ENABLE */
-			struct qspi_cmd cmd = { .op_code = SPI_NOR_CMD_WREN };
+		LOG_DBG("RDSR %02x QE %d need %d: %s", sr, qe_state, qe_value,
+			(qe_state != qe_value) ? "updating" : "no-change");
 
-			if (qspi_send_cmd(dev, &cmd) != 0) {
-				return -EIO;
+		ret = 0;
+		if (qe_state != qe_value) {
+			const struct qspi_buf sr_buf = {
+				.buf = &sr,
+				.len = sizeof(sr),
+			};
+			struct qspi_cmd cmd = {
+				.op_code = SPI_NOR_CMD_WRSR,
+				.tx_buf = &sr_buf,
+			};
+
+			sr ^= qe_mask;
+			ret = qspi_send_cmd(dev, &cmd, true);
+
+			/* Writing SR can take some time, and further
+			 * commands sent while it's happening can be
+			 * corrupted.  Wait.
+			 */
+			if (ret == 0) {
+				ret = qspi_wait_while_writing(dev);
 			}
+		}
+
+		if (ret < 0) {
+			LOG_ERR("QE %s failed: %d", qe_value ? "set" : "clear",
+				ret);
+		}
+	}
 
-			uint8_t tx = BIT(CONFIG_NORDIC_QSPI_NOR_QE_BIT);
+	return ret;
+}
 
-			const struct qspi_buf tx_buff = { .buf = &tx, .len = sizeof(tx), };
+static int qspi_read_jedec_id(const struct device *dev,
+				     uint8_t *id)
+{
+	const struct qspi_buf rx_buf = {
+		.buf = id,
+		.len = 3
```

Comment:
There is identifier `SPI_NOR_MAX_ID_LEN` that represents the length of id and is used in definition of `qspi_nor_config` and here in `qspi_nor_read_id` to define the `id` array that is passed to this function.
Shouldn't the same identifier be used here instead of 3?

Reply:
Fixed this one.